### PR TITLE
Java 21 migration

### PR DESCRIPTION
**Pull request by v-kyrychenko**

### CODE REVIEW

File: xm-commons-ms-web/build.gradle [hunk 1]
```diff
@@ -18,7 +18,8 @@ dependencies {
     }
     implementation "com.fasterxml.jackson.datatype:jackson-datatype-jsr310"
     implementation "com.fasterxml.jackson.datatype:jackson-datatype-hibernate5"
+    implementation "com.fasterxml.jackson.datatype:jackson-datatype-hibernate6"
```
Comment (ivalex0027) [inline xm-commons-ms-web/build.gradle:21]: why we need support both versions?

Comment (v-kyrychenko) [reply xm-commons-ms-web/build.gradle:21]: @ivalex0027 probably not needed,  i will add comment and we will remove it before final commit

Comment (osiievska23) [reply xm-commons-ms-web/build.gradle]: already removed hibernate5

File: xm-commons-tenant/src/main/java/com/icthh/xm/commons/tenant/internal/TenantContextDataHolder.java [hunk 2]
```diff
@@ -28,6 +28,10 @@ final class TenantContextDataHolder {
     private static ThreadLocal<Stack<TenantContextDataHolder>> parentHolderInstanceStack =
         ThreadLocal.withInitial(Stack::new);
 
+    static {
+        ThreadLocalAccessorRegistryInterceptor.register("holderInstance", holderInstance);
+    }
```
Comment (v-kyrychenko) [inline xm-commons-tenant/src/main/java/com/icthh/xm/commons/tenant/internal/TenantContextDataHolder.java:33]: why this changes is requeired?

File: xm-commons-tenant/src/main/java/com/icthh/xm/commons/tenant/internal/ThreadLocalAccessorRegistryInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        try {
+            Class<?> contextRegistry = Class.forName("io.micrometer.context.ContextRegistry");
+            Method getInstance = contextRegistry.getMethod("getInstance");
+            Method registerThreadLocalAccessor = contextRegistry.getMethod("registerThreadLocalAccessor", String.class, ThreadLocal.class);
+
+            Object instance = getInstance.invoke(null);
+            registerThreadLocalAccessor.invoke(instance, key, threadLocal);
+            log.info("Successfully registered ThreadLocal in micrometer ContextRegistry by key '{}'", key);
+
+        } catch (ClassNotFoundException | NoSuchMethodException | LinkageError ignored) {
```
Comment (v-kyrychenko) [inline xm-commons-tenant/src/main/java/com/icthh/xm/commons/tenant/internal/ThreadLocalAccessorRegistryInterceptor.java:28]: why this Exception ignored? maybe at least add log?

File: gradle/lib-versions.gradle [hunk 4]
```diff
@@ -1,50 +1,53 @@
...
-    commonsIo                  : '2.11.0',
+    gradle                     : '8.5',
+    grgit                      : '5.2.1',
+    sonarqube                  : '4.4.1.3373',
+    checkstyle                 : '10.12.5',
+    spotbug                    : '6.0.7',
+    spotbugToolVersion         : '4.8.3',
+    pmd                        : '6.48.0',
+
+    equalsverifier             : '3.10', // 3.15.6
```
[the remark this answers is not in the document]
Comment (osiievska23) [reply gradle/lib-versions.gradle:10]: У новій версії тест падає з помилкою:
```
(EqualsVerifier 3.16, JDK 17.0.12 on Linux)
java.lang.AssertionError: EqualsVerifier found a problem in class com.icthh.xm.commons.tenant.TenantKey.
-> String equality: class uses equalsIgnoreCase to compare String field value, but hashCode is case-sensitive. Use toUpperCase() to determine the hashCode.

```
Для виправлення необхідно оновити метод `hashCode()`. Оскільки `equals()` робить перевірку `equalsIgnoreCase()`, 'ONE' і 'one' будуть еквівалентні, а hashCode буде різний

```
    @Override
    public int hashCode() {
        return Objects.hash(StringUtils.upperCase(value));
    }
```

[the remark this answers is not in the document]
Comment (osiievska23) [reply gradle/lib-versions.gradle]: done

File: gradle/lib-versions.gradle [hunk 5]
```diff
@@ -1,50 +1,53 @@
...
+    commonsText                : '1.11.0',
+    guava                      : '33.0.0-jre',
 
-    springBoot                 : '2.6.6',
-    springCloud                : '2021.0.3',
-    springSecurityOAuth2       : '2.3.7.RELEASE', // migrate to simple JWT
-    springSecurityJwt          : '1.0.11.RELEASE', // migrate to simple JWT
+    springBoot                 : '3.3.0',
+    springCloud                : '2023.0.2',
     springMetrics              : '3.1.3', // TODO check is it need and remove this
```
[the remark this answers is not in the document]
Comment (osiievska23) [reply gradle/lib-versions.gradle]: done

File: xm-commons-metric/build.gradle [hunk 6]
```diff
@@ -8,18 +8,18 @@ dependencyManagement {
...
     api 'org.springframework.kafka:spring-kafka'
     implementation "org.springframework.boot:spring-boot-starter-actuator"
 
     compileOnly "tech.jhipster:jhipster-framework:${versions.jhipster}"
     api "org.apache.commons:commons-lang3:${versions.commonsLang3}"
 
-    api "io.prometheus:simpleclient_dropwizard:0.8.1"
+    api 'io.prometheus:simpleclient_dropwizard:0.16.0'
     api "io.micrometer:micrometer-registry-prometheus"
+    implementation 'io.micrometer:micrometer-registry-prometheus-simpleclient:1.13.0'
```
[the remark this answers is not in the document]
Comment (osiievska23) [reply xm-commons-metric/build.gradle]: done

File: xm-commons-metric/build.gradle [hunk 7]
```diff
@@ -8,18 +8,18 @@ dependencyManagement {
...
+    api "io.dropwizard.metrics:metrics-jvm:${versions.dropwizard}"
     api "com.ryantenney.metrics:metrics-spring:${versions.springMetrics}"
     api 'org.springframework.kafka:spring-kafka'
     implementation "org.springframework.boot:spring-boot-starter-actuator"
 
     compileOnly "tech.jhipster:jhipster-framework:${versions.jhipster}"
     api "org.apache.commons:commons-lang3:${versions.commonsLang3}"
 
-    api "io.prometheus:simpleclient_dropwizard:0.8.1"
+    api 'io.prometheus:simpleclient_dropwizard:0.16.0'
```
[the remark this answers is not in the document]
Comment (osiievska23) [reply xm-commons-metric/build.gradle]: done

File: xm-commons-ms-web/src/main/java/com/icthh/xm/commons/web/spring/TenantInterceptor.java [hunk 8]
```diff
@@ -8,23 +8,23 @@
...
 import java.util.Objects;
 import java.util.Optional;
-import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
 
 /**
  * The {@link TenantInterceptor} class.
  */
 // Add exclusion path to config:
 // https://docs.spring.io/spring/docs/3.2.x/spring-framework-reference/html/mvc.html#mvc-config-interceptors
```
[the remark this answers is not in the document]
Comment (osiievska23) [reply xm-commons-ms-web/src/main/java/com/icthh/xm/commons/web/spring/TenantInterceptor.java]: done

File: xm-commons-scheduler/src/main/java/com/icthh/xm/commons/scheduler/service/SchedulerEventHandlerFacade.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Override
+    public void onMessage(String message, String tenant, TopicConfig topicConfig) {
+        try {
+            MdcUtils.putRid(MdcUtils.generateRid() + ":" + tenant);
+            StopWatch stopWatch = StopWatch.createStarted();
+            String payloadString = unwrap(message);
+            log.debug("start processing message for tenant: [{}], raw body in base64 = {}", tenant, payloadString);
+            String eventBody = new String(Base64.getDecoder().decode(payloadString), UTF_8);
+            if (!eventBody.contains("TEST_2")) {
```
[the remark this answers is not in the document]
Comment (osiievska23) [reply xm-commons-scheduler/src/main/java/com/icthh/xm/commons/scheduler/service/SchedulerEventHandlerFacade.java]: done

File: xm-commons-tenant/src/main/java/com/icthh/xm/commons/tenant/internal/ThreadLocalAccessorRegistryInterceptor.java (hunk 3 above)
Comment (osiievska23) [reply xm-commons-tenant/src/main/java/com/icthh/xm/commons/tenant/internal/ThreadLocalAccessorRegistryInterceptor.java:28]: done

File: xm-commons-tenant/src/main/java/com/icthh/xm/commons/tenant/internal/TenantContextDataHolder.java (hunk 2 above)
Comment (osiievska23) [reply xm-commons-tenant/src/main/java/com/icthh/xm/commons/tenant/internal/TenantContextDataHolder.java:33]: This changes was added to propagate `ThreadLocal<TenantContextDataHolder> holderInstance` between threads in reactive projects. `ThreadLocalAccessorRegistryInterceptor` registers threadLocal in micrometer context registry.

If the thread is registered, the application only needs to enable mdc context propagation by `Hooks.enableAutomaticContextPropagation();`

You could check this related conversation also: [https://github.com/xm-online/xm-commons/pull/245#discussion_r1472843375]

File: xm-commons-logging-trace/src/main/java/com/icthh/xm/commons/logging/trace/TraceService.java [hunk 10]
```diff
@@ -2,22 +2,18 @@
...
 import org.springframework.stereotype.Component;
 
 import static com.icthh.xm.commons.logging.trace.TraceService.TraceServiceField.FIELD_NAME;
 
+/**
+ * @deprecated org.springframework.cloud.sleuth.instrument.messaging.TracingChannelInterceptor is not supported
+ * since Spring Boot 3.x
+ */
+@Deprecated(forRemoval = true)
 @Component
```
[the remark this answers is not in the document]
Comment (v-kyrychenko) [reply xm-commons-logging-trace/src/main/java/com/icthh/xm/commons/logging/trace/TraceService.java:14]: Should be removed. Use     def restTemplate = lepContext.templates.plainRest instead.

[the remark this answers is not in the document]
Comment (osiievska23) [reply xm-commons-logging-trace/src/main/java/com/icthh/xm/commons/logging/trace/TraceService.java]: done

File: xm-commons-migration-db/build.gradle [hunk 11]
```diff
@@ -15,9 +15,7 @@ dependencies {
     api 'org.liquibase:liquibase-core'
     api 'org.springframework:spring-context'
     api 'org.springframework:spring-orm'
-    api 'org.hibernate:hibernate-core'
-
-    implementation "com.vladmihalcea:hibernate-types-52:${versions.hibernateTypes52}"
```
[the remark this answers is not in the document]
Comment (osiievska23) [reply xm-commons-migration-db/build.gradle:20]: banner is missing

File: xm-commons-permission/src/main/java/com/icthh/xm/commons/permission/service/FilterConverter.java [hunk 12]
```diff
@@ -46,7 +46,7 @@ private static Stream<Expression> resolveExpression(Map.Entry<String, Filter> en
         Stream.Builder<Expression> expressions = Stream.builder();
 
         Filter<?> filter = entry.getValue();
-        String fieldName = preProcessForeignKeyField(entry.getKey());
+        String fieldName = preProcessForeignKeyFieldToJpql(entry.getKey()); // fixme: to check
```
[the remark this answers is not in the document]
Comment (osiievska23) [reply xm-commons-permission/src/main/java/com/icthh/xm/commons/permission/service/FilterConverter.java]: done

**Review by amedvedchuk — APPROVED** [review]

**Review by sergeysenja1992 — APPROVED** [review]
